# Decision Trees

Like SVMs, Decision Trees are versatile Machine Learning algorithms that can per
form both classification and regression tasks, and even multioutput tasks. They are
 powerful algorithms, capable of fitting complex datasets.

Decision Trees are also the fundamental components of Random Forests

## Training and Visualizing a Decision Tree

The following code trains a DecisionTreeClassifier on the iris dataset 

In [1]:
from sklearn.datasets import load_iris
from sklearn.tree import DecisionTreeClassifier

In [2]:
iris = load_iris()
X = iris.data[:, 2:] # petal length and width
y = iris.target

In [3]:
X

array([[1.4, 0.2],
       [1.4, 0.2],
       [1.3, 0.2],
       [1.5, 0.2],
       [1.4, 0.2],
       [1.7, 0.4],
       [1.4, 0.3],
       [1.5, 0.2],
       [1.4, 0.2],
       [1.5, 0.1],
       [1.5, 0.2],
       [1.6, 0.2],
       [1.4, 0.1],
       [1.1, 0.1],
       [1.2, 0.2],
       [1.5, 0.4],
       [1.3, 0.4],
       [1.4, 0.3],
       [1.7, 0.3],
       [1.5, 0.3],
       [1.7, 0.2],
       [1.5, 0.4],
       [1. , 0.2],
       [1.7, 0.5],
       [1.9, 0.2],
       [1.6, 0.2],
       [1.6, 0.4],
       [1.5, 0.2],
       [1.4, 0.2],
       [1.6, 0.2],
       [1.6, 0.2],
       [1.5, 0.4],
       [1.5, 0.1],
       [1.4, 0.2],
       [1.5, 0.2],
       [1.2, 0.2],
       [1.3, 0.2],
       [1.4, 0.1],
       [1.3, 0.2],
       [1.5, 0.2],
       [1.3, 0.3],
       [1.3, 0.3],
       [1.3, 0.2],
       [1.6, 0.6],
       [1.9, 0.4],
       [1.4, 0.3],
       [1.6, 0.2],
       [1.4, 0.2],
       [1.5, 0.2],
       [1.4, 0.2],
       [4.7, 1.4],
       [4.5, 1.5],
       [4.9,

In [4]:
y

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

In [5]:
tree_clf = DecisionTreeClassifier(max_depth=4) # IMP!!!! max_depth defines the number of splits the dataset will be divided into, or basically the TOTAL number of leaf nodes from the root willbe there
tree_clf.fit(X, y)

DecisionTreeClassifier(max_depth=4)

In [6]:
iris

{'data': array([[5.1, 3.5, 1.4, 0.2],
        [4.9, 3. , 1.4, 0.2],
        [4.7, 3.2, 1.3, 0.2],
        [4.6, 3.1, 1.5, 0.2],
        [5. , 3.6, 1.4, 0.2],
        [5.4, 3.9, 1.7, 0.4],
        [4.6, 3.4, 1.4, 0.3],
        [5. , 3.4, 1.5, 0.2],
        [4.4, 2.9, 1.4, 0.2],
        [4.9, 3.1, 1.5, 0.1],
        [5.4, 3.7, 1.5, 0.2],
        [4.8, 3.4, 1.6, 0.2],
        [4.8, 3. , 1.4, 0.1],
        [4.3, 3. , 1.1, 0.1],
        [5.8, 4. , 1.2, 0.2],
        [5.7, 4.4, 1.5, 0.4],
        [5.4, 3.9, 1.3, 0.4],
        [5.1, 3.5, 1.4, 0.3],
        [5.7, 3.8, 1.7, 0.3],
        [5.1, 3.8, 1.5, 0.3],
        [5.4, 3.4, 1.7, 0.2],
        [5.1, 3.7, 1.5, 0.4],
        [4.6, 3.6, 1. , 0.2],
        [5.1, 3.3, 1.7, 0.5],
        [4.8, 3.4, 1.9, 0.2],
        [5. , 3. , 1.6, 0.2],
        [5. , 3.4, 1.6, 0.4],
        [5.2, 3.5, 1.5, 0.2],
        [5.2, 3.4, 1.4, 0.2],
        [4.7, 3.2, 1.6, 0.2],
        [4.8, 3.1, 1.6, 0.2],
        [5.4, 3.4, 1.5, 0.4],
        [5.2, 4.1, 1.5, 0.1],
  

In [7]:
type(iris)

sklearn.utils._bunch.Bunch

In [8]:
from sklearn.tree import export_graphviz

In [9]:
image_path = "C:/Users/manas/Downloads/"

In [10]:
 export_graphviz(
 tree_clf,
 out_file="C:/Users/manas/Downloads/iris_tree_dep4.dot",
 feature_names=iris.feature_names[2:],
 class_names=iris.target_names,
 rounded=True,
 filled=True
 )

Then you can use the dot command-line tool from the Graphviz package to convert
 this .dot file to a variety of formats, such as PDF or PNG.1 This command line con
verts the .dot file to a .png image file:

 $ dot -Tpng iris_tree.dot -o iris_tree.png

## Making Predictions

Let’s see how the tree represented makes predictions Suppose you find
 an iris flower and you want to classify it. You start at the root node (depth 0, at the
 top): this node asks whether the flower’s petal length is smaller than 2.45 cm. If it is,
 then you move down to the root’s left child node (depth 1, left). In this case, it is a l node (i.e., it does not have any child nodes), so it does not ask any questions: simply
 look at the predicted class for that node, and the Decision Tree predicts that your
 flower is an Iris setosa (class=setosa).eaf

Now suppose you find another flower, and this time the petal length is greater than
 2.45 cm. You must move down to the root’s right child node (depth 1, right), which is
 not a leaf node, so the node asks another question: is the petal width smaller than
 1.75 cm? If it is, then your flower is most likely an Iris versicolor (depth 2, left). If not,
 it is likely an Iris virginica (depth 2, right). It’s really that simple

### Important point in this cell
One of the many qualities of Decision Trees is that they require
 very little data preparation. In fact, they don’t require feature scal
ing or centering at all.

### Important point in this cell
A node’s samples attribute counts how many training instances it applies to. For
 example, 100 training instances have a petal length greater than 2.45 cm (depth 1,
 right), and of those 100, 54 have a petal width smaller than 1.75 cm (depth 2, left). A
 node’s value attribute tells you how many training instances of each class this node
 applies to: for example, the bottom-right node applies to 0 Iris setosa, 1 Iris versicolor,
 and 45 Iris virginica. Finally, a node’s gini attribute measures its impurity: a node is
 “pure” (gini=0) if all training instances it applies to belong to the same class. For
 example, since the depth-1 left node applies only to Iris setosa training instances, it is
 pure and its gini score is 0.

### Important point in this cell

Scikit-Learn uses the CART algorithm, which produces only binary
 trees: nonleaf nodes always have two children (i.e., questions only
 have yes/no answers). However, other algorithms such as ID3 can
 produce Decision Trees with nodes that have more than two
 children.

## Model Interpretation: White Box Versus Black Box
Decision Trees are intuitive, and their decisions are easy to interpret. Such models are
 often called white box models. In contrast, as we will see, Random Forests or neural
 networks are generally considered black box models. They make great predictions,
 and you can easily check the calculations that they performed to make these predic
tions; nevertheless, it is usually hard to explain in simple terms why the predictions
 were made. For example, if a neural network says that a particular person appears on
 a picture, it is hard to know what contributed to this prediction: did the model recog
nize that person’s eyes? Their mouth? Their nose? Their shoes? Or even the couch
 that they were sitting on? Conversely, Decision Trees provide nice, simple classifica
tion rules that can even be applied manually if need be (e.g., for flower classification).

##  Estimating Class Probabilities

A Decision Tree can also estimate the probability that an instance belongs to a partic
ular class k. First it traverses the tree to find the leaf node for this instance, and then it
 returns the ratio of training instances of class k in this node. For example, suppose
 you have found a flower whose petals are 5 cm long and 1.5 cm wide.  corresponding leaf node is the depth-2 left node, so the Decision Tree should output
 the following probabilities: 0% for Iris setosa (0/54), 90.7% for Iris versicolor (49/54),
 and 9.3% for Iris virginica (5/54). And if you ask it to predict the class, it should out
put Iris versicolor (class 1) because it has the highest probability. Let’s check this:The

In [11]:
tree_clf.predict_proba([[5, 1.5]])

array([[0., 0., 1.]])

In [12]:
tree_clf.predict([[5, 1.5]])

array([2])

## The CART Training Algorithm

Scikit-Learn uses the Classification and Regression Tree (CART) algorithm to train
 Decision Trees (also called “growing” trees). The algorithm works by first splitting the
 training set into two subsets using a single feature k and a threshold tk
 (e.g., “petal
 length ≤ 2.45 cm”). How does it choose k and tk
 ? It searches for the pair (k, tk
 ) that
 produces the purest subsets (weighted by their size). 

Once the CART algorithm has successfully split the training set in two, it splits the
 subsets using the same logic, then the sub-subsets, and so on, recursively. It stops
 recursing once it reaches the maximum depth (defined by the max_depth hyperpara
meter), or if it cannot find a split that will reduce impurity. 

### Warning
As you can see, the CART algorithm is a greedy algorithm: it greed
ily searches for an optimum split at the top level, then repeats the
 process at each subsequent level. It does not check whether or not
 the split will lead to the lowest possible impurity several levels
 down. A greedy algorithm often produces a solution that’s reasona
bly good but not guaranteed to be optimal.
 Unfortunately, finding the optimal tree is known to be an NP
Complete problem:2 it requires O(exp(m)) time, making the prob
lem intractable even for small training sets. This is why we must
 settle for a “reasonably good” solution.

##  Gini Impurity or Entropy?

By default, the Gini impurity measure is used, but you can select the entropy impurity
 measure instead by setting the criterion hyperparameter to "entropy". The concept
 of entropy originated in thermodynamics as a measure of molecular disorder:
 entropy approaches zero when molecules are still and well ordered. Entropy later
 spread to a wide variety of domains, including Shannon’s information theory, where it
 measures the average information content of a message:4 entropy is zero when all
 messages are identical. In Machine Learning, entropy is frequently used impurity measure: a set’s entropy is zero when it contains instances of only one class. So, should you use Gini impurity or entropy? The truth is, most of the time it does
 not make a big difference: they lead to similar trees. Gini impurity is slightly faster to
 compute, so it is a good default. However, when they differ, Gini impurity tends to
 isolate the most frequent class in its own branch of the tree, while entropy tends to
 produce slightly more balanced trees.5 as an

## Regularization Hyperparameters

Decision Trees make very few assumptions about the training data (as opposed to lin
ear models, which assume that the data is linear, for example). If left unconstrained,
 the tree structure will adapt itself to the training data, fitting it very closely—indeed,
 most likely overfitting it. Such a model is often called a nonparametric model, not
 because it does not have any parameters (it often has a lot) but because the number of
 parameters is not determined prior to training, so the model structure is free to stick
 closely to the data. In contrast, a parametric model, such as a linear model, has a pre
determined number of parameters, so its degree of freedom is limited, reducing the
 risk of overfitting (but increasing the risk of underfitting).

To avoid overfitting the training data, you need to restrict the Decision Tree’s freedom
 during training. As you know by now, this is called regularization. The regularization
 hyperparameters depend on the algorithm used, but generally you can at least restrict
 the maximum depth of the Decision Tree. In Scikit-Learn, this is controlled by the
 max_depth hyperparameter (the default value is None, which means unlimited).
 Reducing max_depth will regularize the model and thus reduce the risk of overfitting.

The DecisionTreeClassifier class has a few other parameters that similarly restrict
 the shape of the Decision Tree: min_samples_split (the minimum number of sam
ples a node must have before it can be split), min_samples_leaf (the minimum num
ber of samples a leaf node must have), min_weight_fraction_leaf (same min_samples_leaf but expressed as a fraction of the total number of weighted
 instances), max_leaf_nodes (the maximum number of leaf nodes), and max_features
 (the maximum number of features that are evaluated for splitting at each node).del. as

### Important point in the cell
Increasing min_* hyperparameters or reducing max_* hyperparameters will regularize
 the model.

### Important point in this cell
Other algorithms work by first training the Decision Tree without
 restrictions, then pruning (deleting) unnecessary nodes. A node
 whose children are all leaf nodes is considered unnecessary if the
 purity improvement it provides is not statistically significant. Stan
dard statistical tests, such as the χ2 test (chi-squared test), are used
 to estimate the probability that the improvement is purely the
 result of chance (which is called the null hypothesis). If this proba
bility, called the p-value, is higher than a given threshold (typically
 5%, controlled by a hyperparameter), then the node is considered
 unnecessary and its children are deleted. The pruning continues
 until all unnecessary nodes have been pruned.

## Regression

Decision Trees are also capable of performing regression tasks. Let’s build a regres
sion tree using Scikit-Learn’s DecisionTreeRegressor class, training it on a noisy
 quadratic dataset with max_depth=2:

In [13]:
from sklearn.tree import DecisionTreeRegressor

In [14]:
tree_reg = DecisionTreeRegressor(max_depth=2)
tree_reg.fit(X, y)

DecisionTreeRegressor(max_depth=2)

In [15]:
 export_graphviz(
 tree_reg,
 out_file="C:/Users/manas/Downloads/iris_tree_regression.dot",
 feature_names=iris.feature_names[2:],
 class_names=iris.target_names,
 rounded=True,
 filled=True
 )

### Important point in this cell
The CART algorithm works mostly the same way as earlier, except that instead of try
ing to split the training set in a way that minimizes impurity, it now tries to split the
 training set in a way that minimizes the MSE.

### WARNING
Just like for classification tasks, Decision Trees are prone to overfitting when dealing
 with regression tasks. 

## Instability

Hopefully by now you are convinced that Decision Trees have a lot going for them:
 they are simple to understand and interpret, easy to use, versatile, and powerful.
 However, they do have a few limitations. First, as you may have noticed, Decision
 Trees love orthogonal decision boundaries (all splits are perpendicular to an axis),
 which makes them sensitive to training set rotation.

 More generally, the main issue with Decision Trees is that they are very sensitive to
 small variations in the training data.or example, if you just remove the widest Iris
 versicolor from the iris training set (the one with petals 4.8 cm long and 1.8 cm wide)
 and train a new Decision Tree, you may get the model represented in Figure 6-8.  